src/messages/mod.rs

```diff
@@ -78,19 +78,50 @@ impl HopMessage {
     }
 }
 
+/// Metadata needed for verification of the sender
+#[derive(Ord, PartialOrd, Eq, PartialEq, Clone, Hash, Serialize, Deserialize)]
+pub struct FullSecurityMetadata {
+    proof: SectionProofChain,
+    sender_prefix: Prefix<XorName>,
+    signature: BlsSignature,
+}
+
+impl FullSecurityMetadata {
+    pub fn verify_sig(&self, bytes: Vec<u8>) -> bool {
+        self.proof.last_public_key().verify(&self.signature, bytes)
+    }
+
+    pub fn validate_proof(&self) -> bool {
+        self.proof.validate()
+    }
+
+    pub fn prefix(&self) -> &Prefix<XorName> {
+        &self.sender_prefix
+    }
+
+    pub fn proof_chain(&self) -> &SectionProofChain {
+        &self.proof
+    }
+}
+
+#[derive(Ord, PartialOrd, Eq, PartialEq, Clone, Hash, Serialize, Deserialize)]
+#[allow(clippy::large_enum_variant)]
+pub enum SecurityMetadata {
+    None,
+    Partial(BTreeMap<BlsPublicKeyShare, BlsSignatureShare>),
+    Signed(BlsSignature),
```

Comment:
Do we really need that variant?

Reply:
It's useful because of the thing I mentioned in another comment - that there is no reasonable "empty `SectionProofChain`" value. Thus, it works best if we have a variant that doesn't need to contain such a chain for the time when we have the signature, but haven't attached a chain yet.

Reply:
Related to other comments, I think we probably what to always have the populated SectionProofChain (even in partial).

Reply:
Hmm... Yeah, that's also an option, and would fix some of the problems. Maybe it is a good idea.